---
title: "MAG Generation"
date: 2024-09-24
license: "Apache 2.0"
order: 3.5
author:
  - name: Tatiana Gurbich
    orcid: 0000-0002-7537-8871
    email: tgurbich@ebi.ac.uk
    affiliation: 
      - name: EMBL-EBI
        url: www.ebi.ac.uk

  - name: Ekaterina Sakharova
    orcid: 0000-0001-6867-2039
    email: kates@ebi.ac.uk
    affiliation:
      - name: EMBL-EBI
        url: www.ebi.ac.uk
---

# MAG generation

- Generation of metagenome assembled genomes (MAGs) from assemblies
- Assessment of quality
- Taxonomic assignment

Prerequisites
---------------

For this tutorial, you will need to first start the docker container by running the following command in the terminal:

```bash
sudo docker run --rm -it -v /home/training/Binning:/opt/data quay.io/microbiome-informatics/mags-practical-2024:v2
```
*password: training*

Generating metagenome assembled genomes
----------------------------------------

::: {.callout-note}
Learning Objectives - in the following exercises, you will
learn how to bin an assembly, assess the quality of this assembly with
checkM, and then visualize a placement of these genomes within a
reference tree.
:::

::: {.callout-note}
As with the assembly process, there are many software tools available for
binning metagenome assemblies. Examples include, but are not limited to:

- [MaxBin](https://sourceforge.net/projects/maxbin/)
- [CONCOCT](https://github.com/BinPro/CONCOCT)
- [COCACOLA](https://github.com/younglululu/COCACOLA)
- [MetaBAT](https://bitbucket.org/berkeleylab/metabat)

There is no clear winner between these tools, so it is best to
experiment and compare a few different ones to determine which works
best for your dataset. For this exercise, we will be using **MetaBAT**
(specifically, MetaBAT2). The way in which MetaBAT bins contigs together
is summarized in Figure 1.
:::

![Figure 1. MetaBAT workflow (Kang, et al. *PeerJ* 2015).](mags/binning.png)

### Preparing to run MetaBAT

:::{.callout-step .callout-tip}
Prior to running MetaBAT, we need to generate coverage
statistics by mapping reads to the contigs. To do this, we can use [bwa](http://bio-bwa.sourceforge.net/) and then
[the samtools software](http://www.htslib.org/) to reformat the
output. This can take some time, so we have run it in advance.
:::

:::{.callout-step .callout-tip}
Let's browse the files that we have prepared:
```bash
cd /opt/data/assemblies/
ls
```

You should find the following files in this directory:

- `contigs.fasta`: *a file containing the primary metagenome assembly produced by metaSPAdes (contigs that haven't been binned)*
- `input.fastq.sam.bam`: *a pre-generated file that contains reads mapped back to contigs*

:::

::: {.callout-tip}
If you wanted to generate the `input.fastq.sam.bam` file yourself, you would run the following commands:

```bash

# NOTE: you will not be able to run subsequent steps until this workflow is completed because you need
# the input.fastq.sam.bam file to calculate contig depth in the next step. In the interest of time, we
# suggest that you try the commands after you complete the practical if you would like to.

# If you would like to practice this now, back up the input.fastq.sam.bam file that we provided first,
# as these steps will take a while:

cd /opt/data/assemblies/
mv input.fastq.sam.bam input.fastq.sam.bam.bak

# index the contigs file that was produced by metaSPAdes:
bwa index contigs.fasta

# fetch the reads from ENA
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR011/ERR011322/ERR011322_1.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR011/ERR011322/ERR011322_2.fastq.gz

# map the original reads to the contigs:
bwa mem contigs.fasta ERR011322_1.fastq ERR011322_2.fastq > input.fastq.sam

# reformat the file with samtools:
samtools view -Sbu input.fastq.sam > junk 
samtools sort junk -o input.fastq.sam.bam
```
:::

### Running MetaBAT

:::{.callout-step .callout-tip}
# Create a subdirectory where files will be saved to


```bash
cd /opt/data/assemblies/
mkdir contigs.fasta.metabat-bins2000
```
:::

In this case, the directory might already be part of your VM, so do not worry if you get an error saying the directory already exists. You can move on to the next step.


:::{.callout-step .callout-tip}
Run the following command to produce a
`contigs.fasta.depth.txt` file, summarizing the output depth for use with
MetaBAT:

```bash
jgi_summarize_bam_contig_depths --outputDepth contigs.fasta.depth.txt input.fastq.sam.bam
```
:::

:::{.callout-step .callout-tip}
Now let's put together the metaBAT2 command. To see the available options, run:

```bash
metabat2 -h
```

Here is what we are trying to do:  

* we want to bin the assembly file called `contigs.fasta`

* the resulting bins should be saved into the `contigs.fasta.metabat-bins2000` folder

* we want the bin file names to start with the prefix `bin`

* we want to use the contig depth file we just generated (`contigs.fasta.depth.txt`)

* the minimum contig length should be 2000

Take a moment to put your command together but please check the answer below before running it to make sure everything is correct.


<details>

<summary>See the answer</summary>

```bash
metabat2 --inFile contigs.fasta --outFile contigs.fasta.metabat-bins2000/bin --abdFile contigs.fasta.depth.txt --minContig 2000
```
</details>

:::

:::{.callout-note}
Once the binning process is complete, each bin will be
grouped into a multi-fasta file with a name structure of
**bin.[0-9].fa**.
:::

:::{.callout-step .callout-tip}
Inspect the output of the binning process.

```bash
ls contigs.fasta.metabat-bins2000/bin*
```
:::

:::{.callout-question .callout-tip}
How many bins did the process produce?
:::

:::{.callout-question .callout-tip}
How many sequences are in each bin?
:::

Obviously, not all bins will have the same level of accuracy since some
might represent a very small fraction of a potential species present in
your dataset. To further assess the quality of the bins, we will use
[**CheckM**](https://github.com/Ecogenomics/CheckM/wiki).

### Running CheckM

:::{.callout-note}
**CheckM** has its own reference database of single-copy
marker genes. Essentially, based on the proportion of these markers
detected in the bin, the number of copies of each, and how different they
are, it will determine the level of **completeness**, **contamination**,
and **strain heterogeneity** of the predicted genome.
:::

:::{.callout-step .callout-tip}
Before we start, we need to configure CheckM.

```bash
cd /opt/data
mkdir /opt/data/checkm_data
tar -xf checkm_data_2015_01_16.tar.gz -C /opt/data/checkm_data
checkm data setRoot /opt/data/checkm_data
```
:::


This program has some handy tools not only for quality control but also
for taxonomic classification, assessing coverage, building a
phylogenetic tree, etc. The most relevant ones for this exercise are
wrapped into the **lineage_wf** workflow.


:::{.callout-step .callout-tip}
STOP!!!! The following command fails. Please use backups instead, inspect the command visually only:

```bash
cd /opt/data/assemblies
# INSPECT BUT DO NOT RUN!
checkm lineage_wf -x fa contigs.fasta.metabat-bins2000 checkm_output --tab_table -f MAGs_checkm.tab --reduced_tree -t 4
```

Due to memory constraints (< 40 GB), we have added the option
`--reduced_tree` to build the phylogeny with a reduced number of
reference genomes.

Once the **lineage_wf** analysis is done, the reference tree can be
found in **checkm_output/storage/tree/concatenated.tre**. 

Additionally, you will have the taxonomic assignment and quality assessment of each
bin in the file **MAGs_checkm.tab** with the corresponding level of
**completeness**, **contamination**, and **strain heterogeneity** (Fig.
2). A quick way to infer the overall quality of the bin is to calculate
the level of **(completeness - 5*contamination)**. You should be aiming for an overall score of at
least **70-80%**.

You can inspect the CheckM output with:

```bash
cat MAGs_checkm.tab
```
:::


![Example output of CheckM](mags/checkm.png)

### Comparing Checkm and Checkm2

Today researchers also use Checkm2, an improved method of predicting genome quality of MAGs using machine learning.
The execution of Checkm2 takes more time than Checkm. We have pre-generated quality TSV table.
It is available on our FTP.

```bash
# create folder for checkm2 result
cd /opt/data/assemblies
mkdir checkm2
cd checkm2
# download checkm2 TSV result
wget https://ftp.ebi.ac.uk/pub/databases/metagenomics/mgnify_courses/metagenomics_2024/mags/checkm2_quality_report.tsv
```

If you would like to try Checkm2 launch it separately:

```bash
cd /opt/data/assemblies
docker run -t quay.it/biocontainers/checkm2:1.0.2--pyh7cba7a3_0 checkm2 predict -i contigs.fasta.metabat-bins2000 checkm_output -o checkm2_result --database_path uniref100.KO.1.dmnd -x fa --force --lowmem -t 4
```

We can compare Checkm and Checkm2 results using scatter plot. X axis - completeness, Y axis - contamination.
We have created a simple python script to plot each Checkm result.
Input table should be in CSV format and contain 3 columns:
bin name,completeness and contamination.
Required file header: "bin,completeness,contamination"

Modify table and plot Checkm:
```bash
# modify MAGs_checkm.tab
# add header
echo "bin,completeness,contamination" > checkm1_quality_report.csv
# take file without header; leave only columns 1, 12 and 13 ; replace tab to comma
tail -n+2 MAGs_checkm.tab | cut -f1,12,13 | tr '\t' ',' >> checkm1_quality_report.csv

# plot
complenetess_vs_contamination.py -i checkm1_quality_report.csv -o checkm1
```

Modify table and plot Checkm2:
```bash
# create CSV
# add header
echo "bin,completeness,contamination" > checkm2_quality_report.csv
# take file without header, leave only first 3 columns, replace tab to comma
tail -n+2 checkm2_quality_report.tsv | cut -f1-3  | tr '\t' ',' >> checkm2_quality_report.csv

# plot
complenetess_vs_contamination.py -i checkm1_quality_report.csv -o checkm2
```

Now you should get checkm1.png and checkm2.png to compare differences in quality predictions.

:::

## Visualizing the phylogenetic tree

### GTDB-Tk taxonomy

Due to long computational time we have pre-launched GTDB-Tk annotation on all bins and put results to FTP.
Lets have a check assigned taxonomy.

```bash
cd /opt/data/assemblies

# download a table
wget https://ftp.ebi.ac.uk/pub/databases/metagenomics/mgnify_courses/metagenomics_2024/mags/taxonomy/bins_gtdbtk.bac120.summary.tsv

# open table
cat bins_gtdbtk.bac120.summary.tsv
```

:::{.callout-question .callout-tip}
How many bins were assigned to species level?
:::

:::{.callout-question .callout-tip}
Make a guess why some bins were not classified to any pylum?
:::

Now you can check your classified bins on small bacterial phylogenetic tree with iTOL.
iTOL is a quick and user-friendly way to do this is to use the web-based **interactive Tree
of Life** [**iTOL**](http://itol.embl.de/index.shtml)

:::{.callout-step .callout-tip}
To use **iTOL** you will need a user account. For the
purpose of this tutorial we have already created one for you with an
example tree.
:::

#### Using EBI account

Go to the [**iTOL**](http://itol.embl.de/index.shtml) website (open the link in a new window).

The login is as follows:

**User:**  *EBI_training*

**Password:**  *EBI_training*

After you login, just click on **My Trees** in the toolbar at the top
and select
**mags_gtdb_fasttree_150.nwk** from the **Imported trees** workspace (use might be located on page 3).

#### Using personal account

Alternatively, if you want to create your own account or you already have an account you can plot the tree
yourself follow these steps:

1. Pre-download tree files from FTP
- tree: https://ftp.ebi.ac.uk/pub/databases/metagenomics/mgnify_courses/metagenomics_2024/mags/taxonomy/small_bac_tree.nwk
- legend: https://ftp.ebi.ac.uk/pub/databases/metagenomics/mgnify_courses/metagenomics_2024/mags/taxonomy/small_bac_tree.legend.txt
- layers: https://ftp.ebi.ac.uk/pub/databases/metagenomics/mgnify_courses/metagenomics_2024/mags/taxonomy/small_bac_tree.layer.txt
2. After you have created and logged in to your account go to **My Trees**
3. From there select **Upload tree files** and upload tree file **small_bac_tree.nwk**
4. Once uploaded, click the tree name to visualize the plot
5. To colour the clades and the outside circle according to the phylum of each strain, drag and drop the files **small_bac_tree.legend.txt** and **small_bac_tree.layer.txt**

:::{.callout-question .callout-tip}
Have you found your bins on that tree?
:::

### How to deal with unclassified bins?

Now we can try another method to classify bins using Checkm tree.

Before we can visualize and plot the Checkm tree, we will need to convert the
reference ID names used by CheckM to taxon names. We have already
prepared a mapping file for renaming the tree (**rename_list.tab**). We
can then do this easily with the [**newick utilities**](http://cegg.unige.ch/newick_utils).

:::{.callout-step .callout-tip}
To do this, run the following command:

```bash
cd /opt/data/
nw_rename assemblies/checkm_output/storage/tree/concatenated.tre assemblies/rename_list.tab > renamed.tree
```

Now we will get back to iTOL visualisation.

**iTOL** only takes in newick formatted trees, so we need to quickly
reformat the tree with [**FigTree**](http://tree.bio.ed.ac.uk/software/figtree/).

:::{.callout-step .callout-tip}
In order to open **FigTree**, open a new terminal window (without docker) and type `java -jar /home/training/JAR/figtree.jar`
:::


:::{.callout-step .callout-tip}
Open the **renamed.tree** file with **FigTree** (**File -> Open**) (file is in `/home/training/Binning`) and then
select from the toolbar **File -> Export Trees**. In the **Tree file
format** select **Newick** and export the file as **renamed.nwk** (or choose a name you will recognise if you plan to use the shared account described below).
:::

#### iTOL visualisation

1. Login into your working account (EBI/personal)
2. Go again to **My Trees**
3. Open/upload **IBD_checkm.nwk**
4. [personal usage only ] add clades drag and drop the files **iTOL_clades.txt** and **iTOL_ocircles.txt** present in /home/training/Data/Binning/iTOL_Files/ into the browser window


Once that is done, all the reference genomes used by **CheckM** will be
coloured according to their phylum name, while all the other ones left
blank correspond to the **target genomes** we placed in the tree.
Highlighting each tip of the phylogeny will let you see the whole
taxon/sample name. Feel free to play around with the plot.

To find the bins you generated, click on the search icon in the left-hand side menu (magnifying glass
with the letters "Aa" in it). In the search field type **bin.**. Click on each bin name to see it in the tree.

:::{.callout-question .callout-tip}
Does the CheckM taxonomic classification make sense? Were you able to find all of the bins in the tree? If not, why do you think that is?
:::
