<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Samuel Horsfield">
<meta name="author" content="Johanna von Wachsmann,">
<meta name="dcterms.date" content="2025-09-19">

<title>Pangenome analysis of metagenomic data â€“ Metagenomics bioinformatics at MGnify (2025)</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../sessions/getting_data_from_mgnify.html" rel="prev">
<link href="../static/mgnify_logo.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-64358e5ab03b06226dd34ee249bba1f1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-ca146c75af252fb53c9e73e81359de52.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar docked nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    </a>
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Metagenomics bioinformatics at MGnify (2025)</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="../index.html" aria-current="page"> 
<span class="menu-text">Home</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://www.ebi.ac.uk/metagenomics"> 
<span class="menu-text">Open MGnify</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://www.ebi.ac.uk/training/events/metagenomics-bioinformatics-2025/#vf-tabs__section--tab2"> 
<span class="menu-text">Course Programme</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../sessions/qc.html">Sessions</a></li><li class="breadcrumb-item"><a href="../sessions/pangenome.html">Pangenome analysis of metagenomic data</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
      <a href="../index.html" class="sidebar-logo-link">
      </a>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Sessions</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../sessions/qc.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quality control and filtering of the raw sequence files</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../sessions/assemblies.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assembly and Co-assembly of Metagenomic Raw Reads</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../sessions/mags.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">MAG Generation</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../sessions/genome_uploader.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">The genome_uploader: tutorial</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../sessions/mag_catalogues.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">MAG Catalogues</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../sessions/getting_data_from_mgnify.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Getting data from MGnify</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../sessions/pangenome.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Pangenome analysis of metagenomic data</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#preamble" id="toc-preamble" class="nav-link active" data-scroll-target="#preamble">Preamble</a></li>
  <li><a href="#building-poppunk-models" id="toc-building-poppunk-models" class="nav-link" data-scroll-target="#building-poppunk-models">1. Building PopPUNK models</a>
  <ul class="collapse">
  <li><a href="#creating-the-database" id="toc-creating-the-database" class="nav-link" data-scroll-target="#creating-the-database">Creating the database</a></li>
  <li><a href="#fitting-a-model" id="toc-fitting-a-model" class="nav-link" data-scroll-target="#fitting-a-model">Fitting a model</a></li>
  <li><a href="#visualising-the-results" id="toc-visualising-the-results" class="nav-link" data-scroll-target="#visualising-the-results">Visualising the results</a></li>
  <li><a href="#further-analysis" id="toc-further-analysis" class="nav-link" data-scroll-target="#further-analysis">Further analysis</a></li>
  </ul></li>
  <li><a href="#assigning-using-poppunk-models" id="toc-assigning-using-poppunk-models" class="nav-link" data-scroll-target="#assigning-using-poppunk-models">2. Assigning using PopPUNK models</a>
  <ul class="collapse">
  <li><a href="#using-poppunk_assign-to-assign-mags-to-subspecies" id="toc-using-poppunk_assign-to-assign-mags-to-subspecies" class="nav-link" data-scroll-target="#using-poppunk_assign-to-assign-mags-to-subspecies">Using <code>poppunk_assign</code> to assign MAGs to subspecies</a></li>
  <li><a href="#updating-the-database" id="toc-updating-the-database" class="nav-link" data-scroll-target="#updating-the-database">Updating the database</a></li>
  <li><a href="#further-analysis-1" id="toc-further-analysis-1" class="nav-link" data-scroll-target="#further-analysis-1">Further analysis</a></li>
  </ul></li>
  <li><a href="#finding-core-and-accessory-genes" id="toc-finding-core-and-accessory-genes" class="nav-link" data-scroll-target="#finding-core-and-accessory-genes">3. Finding core and accessory genes</a>
  <ul class="collapse">
  <li><a href="#celebrimbor-a-pipeline-for-pangenome-analysis-and-threshold-correction" id="toc-celebrimbor-a-pipeline-for-pangenome-analysis-and-threshold-correction" class="nav-link" data-scroll-target="#celebrimbor-a-pipeline-for-pangenome-analysis-and-threshold-correction">CELEBRIMBOR: a pipeline for pangenome analysis and threshold correction</a></li>
  <li><a href="#further-analysis-2" id="toc-further-analysis-2" class="nav-link" data-scroll-target="#further-analysis-2">Further analysis</a></li>
  </ul></li>
  <li><a href="#example-analysis" id="toc-example-analysis" class="nav-link" data-scroll-target="#example-analysis">Example analysis</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../sessions/qc.html">Sessions</a></li><li class="breadcrumb-item"><a href="../sessions/pangenome.html">Pangenome analysis of metagenomic data</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">Pangenome analysis of metagenomic data</h1>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Authors</div>
  <div class="quarto-title-meta-heading">Affiliation</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Samuel Horsfield <a href="mailto:shorsfield@ebi.ac.uk" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> <a href="https://orcid.org/0000-0002-3859-4073" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            <a href="https://www.ebi.ac.uk">
            EMBL-EBI
            </a>
          </p>
      </div>
    <div class="quarto-title-meta-contents">
    <p class="author">Johanna von Wachsmann, <a href="mailto:wachsmannj@ebi.ac.uk" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> <a href="https://orcid.org/0000-0002-4464-3900" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            <a href="https://www.ebi.ac.uk">
            EMBL-EBI
            </a>
          </p>
      </div>
  </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">September 19, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="preamble" class="level2">
<h2 class="anchored" data-anchor-id="preamble">Preamble</h2>
<p>We will first run through how to use our tool PopPUNK to build models to find subclusters within isolates data, then how to use these models (or pre-existing models) to assign MAG data to these clusters.</p>
<p>We anticipate the following timings:</p>
<ol type="1">
<li>Building a PopPUNK model â€“ 45 minutes.</li>
<li>Assigning using a PopPUNK model â€“ 30 minutes.</li>
</ol>
<p>Please do not spend significantly longer than this on each one if you wish to complete the practical. You can move onto the next section at any time.</p>
<p>We will then show how to correct MAG data for incompleteness to give each gene a classification of core or accessory.</p>
<ol start="3" type="1">
<li>Finding core and accessory genes from MAG data â€“ 45 minutes.</li>
</ol>
<p>Use the sidebar to see the instructions for each part.</p>
<p>The files for all practicals are available on the virtual machine here:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> ~/course_dir/data_dir/MGnify_training_course</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>The course prerequisities are already install on the VM. To activate the environment, run:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> ~/pangenome_mgnify_env/bin/activate</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>If the enviroment has been found correctly, you should see something similar to the following on your terminal</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="kw">(</span><span class="ex">pangenome_mgnify_env</span><span class="kw">)</span> <span class="ex">training@user:~$</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="building-poppunk-models" class="level2">
<h2 class="anchored" data-anchor-id="building-poppunk-models">1. Building PopPUNK models</h2>
<p>We will be using 112 <em>B. uniformis</em> isolate genomes (i.e.&nbsp;not MAG data). We are going to use these to define subspecies within the population using <a href="https://genome.cshlp.org/content/29/2/304">PopPUNK</a>. These can be listed using:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> ~/course_dir/data_dir/MGnify_training_course/MAGs_to_build_model</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p><a href="https://poppunk-docs.bacpop.org/installation.html">Installation instructions</a> and an <a href="https://poppunk-docs.bacpop.org/overview.html">overview</a> are available.</p>
<p>PopPUNK is already installed on the VM. Confirm this by running <code>poppunk --version</code>.</p>
<section id="creating-the-database" class="level3">
<h3 class="anchored" data-anchor-id="creating-the-database">Creating the database</h3>
<p>The first step to running PopPUNK on a new species is to create a â€˜databaseâ€™ which contains sketches of the genomes and calculates all of the core and accessory distances between the samples. We will be following the guidance in the <a href="https://poppunk-docs.bacpop.org/sketching.html#">relevant section of the documentation</a>.</p>
<p>First, navigate to the working directory and create a new directory:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/course_dir/work_dir/Day_5 <span class="kw">&amp;&amp;</span> <span class="fu">mkdir</span> pangenome <span class="kw">&amp;&amp;</span> <span class="bu">cd</span> pangenome</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>We need to create a list of the input files. This needs to have the sample names and the location of files with their genomes. This can be created in many ways, here we will use a simple bash command:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">paste</span> <span class="op">&lt;(</span><span class="fu">ls</span> ~/course_dir/data_dir/MGnify_training_course/MAGs_to_build_model <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d</span> <span class="st">"."</span> <span class="at">-f</span> 1 <span class="op">)</span> <span class="op">&lt;(</span><span class="fu">ls</span> ~/course_dir/data_dir/MGnify_training_course/MAGs_to_build_model/MGYG<span class="pp">*</span><span class="op">)</span> <span class="op">&gt;</span> rfile.txt</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>We can now create the input database with <code>--create-db</code>.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Challenge
</div>
</div>
<div class="callout-body-container callout-body">
<p>Using the <a href="https://poppunk-docs.bacpop.org/sketching.html#">documentation</a>, create sketches of the fasta files, using a minimum k of 17, and maximum k of 33. Also, plot the fits to determine whether the values of k are suitable. The command will look like:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">poppunk</span> <span class="at">--create-db</span> <span class="at">--output</span> b_uniformis_db <span class="at">--r-files</span> rfile.txt <span class="at">--min-k</span> <span class="op">&lt;</span>replace<span class="op">&gt;</span> --max-k <span class="op">&lt;</span>replace<span class="op">&gt;</span> --plot-fit 10 <span class="at">--threads</span> 4</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center collapsed" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Note</span>Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>To generate sketches and k vs.&nbsp;matching k plots:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">poppunk</span> <span class="at">--create-db</span> <span class="at">--output</span> b_uniformis_db <span class="at">--r-files</span> rfile.txt <span class="at">--min-k</span> 17 <span class="at">--max-k</span> 33 <span class="at">--plot-fit</span> 10 <span class="at">--threads</span> 4</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>This will run on 4 CPU cores to increase the speed. The other option is the range of k-mer lengths used, see the <a href="https://poppunk-docs.bacpop.org/sketching.html#choosing-the-right-k-mer-lengths">documentation</a> for more information.</p>
</div>
</div>
</div>
<p>Have a look at the plots in the output directory e.g.&nbsp;<code>b_uniformis_db/b_uniformis_db_distanceDistribution.png</code> and some of the fit examples such as <code>b_uniformis_db_fit_example_1.pdf</code>.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Challenge
</div>
</div>
<div class="callout-body-container callout-body">
<p>Can you identify clusters by eye? Is the relationship between k and the proportion of matching k-mers linear?</p>
</div>
</div>
<p>Usually we want to run some <a href="https://poppunk-docs.bacpop.org/qc.html">quality control</a> on the data.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Challenge
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try this with a maximum core (pi) distance of 0.02 and a maximum accessory (a) distance of 1. Try also with different values and compare the results. The command will be:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">poppunk</span> <span class="at">--qc-db</span> <span class="at">--ref-db</span> b_uniformis_db <span class="at">--max-pi-dist</span> <span class="op">&lt;</span>replace<span class="op">&gt;</span> --max-a-dist <span class="op">&lt;</span>replace <span class="op">&gt;</span> --output b_uniformis_db_qc</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center collapsed" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Note</span>Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>To run qc:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">poppunk</span> <span class="at">--qc-db</span> <span class="at">--ref-db</span> b_uniformis_db <span class="at">--max-pi-dist</span> 0.02 <span class="at">--max-a-dist</span> 1 <span class="at">--output</span> b_uniformis_db_qc</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>This removes outlying distances and poor quality samples.</p>
<p>However if we run with a smaller core (pi) distance, this will remove half of the samples.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">poppunk</span> <span class="at">--qc-db</span> <span class="at">--ref-db</span> b_uniformis_db <span class="at">--max-pi-dist</span> 0.01 <span class="at">--max-a-dist</span> 1 <span class="at">--output</span> b_uniformis_db_qc_strict</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>This isnâ€™t a good idea here as this core distance is too strict for a species as diverse as <em>B. uniformis</em>.</p>
</div>
</div>
</div>
<p>This is an example of how to condict quality control. In this case the data is all good quality and should all be retained.</p>
</section>
<section id="fitting-a-model" class="level3">
<h3 class="anchored" data-anchor-id="fitting-a-model">Fitting a model</h3>
<p>We now need to create a â€˜modelâ€™ which determines a cutoff below which genomes are considered the same subspecies.</p>
<p>There are many options available, as detailed in the <a href="https://poppunk-docs.bacpop.org/model_fitting.html">documentation</a>.</p>
<p>As we have â€˜smallâ€™ sample collection with strain-structure where distance distribution components are clearly separated, weâ€™ll try the Bayesian Gaussian Mixture Model with two components:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">poppunk</span> <span class="at">--fit-model</span> bgmm <span class="at">--ref-db</span> b_uniformis_db <span class="at">--output</span> b_uniformis_BGMM_K2 <span class="at">--K</span> 2</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>From the output to the console we can see that everything is in one cluster (In <code>Network Summary</code> there is one component) and so we havenâ€™t found any subspecies. Have a look at the output plot <code>b_uniformis_BGMM_K2/b_uniformis_BGMM_K2_DPGMM_fit.png</code> too.</p>
<p>It looks like adding an extra component (â€˜blobâ€™) might help, so letâ€™s try that and make three rather than two.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Challenge
</div>
</div>
<div class="callout-body-container callout-body">
<p>Fit the BGMM model using three components. How does it compare visually to using two components?</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center collapsed" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Note</span>Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>To run with more components:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="ex">poppunk</span> <span class="at">--fit-model</span> bgmm <span class="at">--ref-db</span> b_uniformis_db <span class="at">--output</span> b_uniformis_BGMM_K3 <span class="at">--K</span> 3</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>That looks better and there are some clusters (In <code>Network Summary</code> there are is seven/eight components), but if you look at the same plot again (<code>b_uniformis_BGMM_K3/b_uniformis_BGMM_K3_DPGMM_fit.png</code>) it doesnâ€™t look like a great fit to the data.</p>
</div>
</div>
</div>
<p>In this case the data is actually a bit sparse to automatically get a good fit and we can do a much better job if we enrich the dataset with a few thousand MAGs and then use the â€˜refineâ€™ model mode. But for now weâ€™ll take a shortcut and impose a cutoff that looks like it will work by using the <code>threshold</code> model.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Challenge
</div>
</div>
<div class="callout-body-container callout-body">
<p>Based on the first cluster closest to the origin, come up with a suitable threshold.</p>
<p>The aim is to define a cutoff which separates the cluster closest to the origin from all others. The command will be:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">poppunk</span> <span class="at">--fit-model</span> threshold <span class="at">--ref-db</span> b_uniformis_db <span class="at">--output</span> b_uniformis_threshold <span class="at">--threshold</span> <span class="op">&lt;</span>replace<span class="op">&gt;</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center collapsed" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Note</span>Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Based on where the first cluster closest to the origin is, it seems a threshold of core distance at 0.0025 is the best:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="ex">poppunk</span> <span class="at">--fit-model</span> threshold <span class="at">--ref-db</span> b_uniformis_db <span class="at">--output</span> b_uniformis_threshold <span class="at">--threshold</span> 0.0025</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>That looks better and there are some clusters (In <code>Network Summary</code> there are is seven components), but if you look at the same plot again (<code>b_uniformis_BGMM_K3/b_uniformis_BGMM_K3_DPGMM_fit.png</code>) it doesnâ€™t look like a great fit to the data.</p>
</div>
</div>
</div>
<p>There actually isnâ€™t one â€˜perfectâ€™ cutoff, which is supported by population genetic simulations (see this <a href="https://genome.cshlp.org/content/early/2023/05/30/gr.277395.122">paper</a>). So we are acutally free to choose a cutoff that defines subspecies which â€˜workâ€™ for us, and the models in PopPUNK are mostly a more convenient way of automating this to find useful clusters in general.</p>
<p>As you can see, this is all a bit fiddly and requires iteration. It is usually better to use a previously fitted and tested model, which we will cover in the next part.</p>
</section>
<section id="visualising-the-results" class="level3">
<h3 class="anchored" data-anchor-id="visualising-the-results">Visualising the results</h3>
<p>But before we move on, letâ€™s get a better look at the results. We can make a core genome tree and accessory genome embedding and plot the clusters interactively in the browser. First, run the <a href="https://poppunk-docs.bacpop.org/visualisation.html#"><code>poppunk_visualise</code> command</a>.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="ex">poppunk_visualise</span> <span class="at">--ref-db</span> b_uniformis_db <span class="at">--model-dir</span> b_uniformis_threshold <span class="at">--microreact</span> <span class="at">--maxIter</span> 100000 <span class="at">--output</span> b_uniformis_db_viz</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>Here, <code>maxIter</code> is being used to reduce the number embedding iterations as the dataset is small, just so the command runs quickly.</p>
<p>Now, open up <a href="https://microreact.org/" class="uri">https://microreact.org/</a> in your browser and choose â€˜Uploadâ€™. Drag and drop the <code>.microreact</code> file in the <code>b_uniformis_db_viz</code> directory to see the clusters and the tree. To change the tree layout, select the circular tree option in the top right corner.</p>
<p>The left-hand panel shows clustering by accessory genome distance, whilst the right-hand panel shows the tree generated from core genome distances. The points are coloured by which sequence cluster they were assigned to by PopPUNK.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Challenge
</div>
</div>
<div class="callout-body-container callout-body">
<p>Check whether the cluster assignments (colours of tips/points) matches up with the phylogeny. Are the cluster assignments in agreement with the core genome tree and the accessory distances?</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center collapsed" data-bs-toggle="collapse" data-bs-target=".callout-11-contents" aria-controls="callout-11" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Note</span>Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-11" class="callout-11-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>As we only used a core genome threshold, it is likely that the cluster assignments will be in better aggreement with the core distances rather than the accessory distances. Using different models to identify sequence clusters (e.g.&nbsp;BGMM) will give different results.</p>
</div>
</div>
</div>
</section>
<section id="further-analysis" class="level3">
<h3 class="anchored" data-anchor-id="further-analysis">Further analysis</h3>
<p>If you have time, try playing around with different methods of fitting PopPUNK models in the <a href="https://poppunk-docs.bacpop.org/model_fitting.html#">documentation</a>.</p>
<p>You can experiment with <a href="https://hdbscan.readthedocs.io/en/latest/how_hdbscan_works.html">HDBSCAN</a>, a method for automatically detecting the number and position of clusters.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Challenge
</div>
</div>
<div class="callout-body-container callout-body">
<p>How do the number of identified sequence clusters compare between using BGMM, HDBSCAN and threshold models?</p>
</div>
</div>
<p>Once you have a model fitted using BGMM or HDBSCAN, you can also <a href="https://poppunk-docs.bacpop.org/model_fitting.html#refine">refine</a> it. This method takes a previously identified within-strain boundary and moves it to optimise the network score of the strain cluster network. We always recommend refining a previous fit, as it may significantly improve strain assignments.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Challenge
</div>
</div>
<div class="callout-body-container callout-body">
<p>Again, how do the number of identified sequence clusters compare when using refinement?</p>
</div>
</div>
<p>You can also play around with visualisation tools such as <a href="https://cytoscape.org/what_is_cytoscape.html">Cytoscape</a>. Use the <code>poppunk_visualise</code> tool to generate a cytoscape output following the <a href="https://poppunk-docs.bacpop.org/visualisation.html#cytoscape">documentation</a>. This will allow you to visualise the strain cluster network, the components of which are the strains detected by PopPUNK. Note: this will generate a file for each component, as well as the whole network, enabling visualisation of the whole network or just parts of it. To open the file, open <code>Cytoscape</code> -&gt; <code>File</code> -&gt; <code>Import</code> -&gt; <code>Network from File</code> and them open a <code>.gml</code> file.</p>
</section>
</section>
<section id="assigning-using-poppunk-models" class="level2">
<h2 class="anchored" data-anchor-id="assigning-using-poppunk-models">2. Assigning using PopPUNK models</h2>
<p>It is faster to <a href="https://poppunk-docs.bacpop.org/query_assignment.html">â€˜assignâ€™</a> new samples to an existing definition of subspecies. This has the bonus that their names will be consistent across studies.</p>
<p>Typically, you can download an existing database with a fit from <a href="https://www.bacpop.org/poppunk-databases" class="uri">https://www.bacpop.org/poppunk-databases</a>. If you have a fit for a new species please send it to us and we can share it here.</p>
<p>There is no fit for <em>B. uniformis</em> (yetâ€¦) so weâ€™ll use the one we just made.</p>
<section id="using-poppunk_assign-to-assign-mags-to-subspecies" class="level3">
<h3 class="anchored" data-anchor-id="using-poppunk_assign-to-assign-mags-to-subspecies">Using <code>poppunk_assign</code> to assign MAGs to subspecies</h3>
<p>Now weâ€™ll work with a large collection of MAGs. These are the <em>B. uniformis</em> MAGs from MGnify with &gt;95% completeness and &lt;5% contamination. They can be listed here</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> ~/course_dir/data_dir/MGnify_training_course/MAGs_to_assign</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>PopPUNK distances are relatively robust to missing sequence content seen in MAGs, but less able to deal with contamination.</p>
<p>Again, navigate to the work directory and create the list of input files for poppunk:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/course_dir/work_dir/Day_5/pangenome</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">paste</span> <span class="op">&lt;(</span><span class="fu">ls</span> ~/course_dir/data_dir/MGnify_training_course/MAGs_to_assign <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-d</span> <span class="st">"."</span> <span class="at">-f</span> 1 <span class="op">)</span> <span class="op">&lt;(</span><span class="fu">ls</span> ~/course_dir/data_dir/MGnify_training_course/MAGs_to_assign/<span class="pp">*</span>.fasta<span class="op">)</span> <span class="op">&gt;</span> qfile.txt</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>One small problem is that these MAGs also contain the isolates from before. PopPUNK will refuse to assign these without unique names. Hereâ€™s a bash incantation to remove the duplicates:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb19"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f</span> 1 rfile.txt <span class="op">&gt;</span> ref_names.txt <span class="kw">&amp;&amp;</span> <span class="fu">grep</span> <span class="at">-v</span> <span class="at">-w</span> <span class="at">-f</span> ref_names.txt qfile.txt <span class="op">&gt;</span> qfile_nodups.txt</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> qfile_nodups.txt qfile.txt</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>The command is relatively simple, we need to provide the database, the model directory and the input â€˜queryâ€™ genomes to assign. Quality control is â€˜built-inâ€™:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb20"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="ex">poppunk_assign</span> <span class="at">--db</span> b_uniformis_db <span class="at">--model-dir</span> b_uniformis_threshold <span class="at">--query</span> qfile.txt <span class="at">--output</span> b_uniformis_queries <span class="at">--threads</span> 4 <span class="at">--max-merge</span> 3 <span class="at">--run-qc</span> <span class="at">--max-a-dist</span> 0.8</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>The <code>b_uniformis_queries_clusters.csv</code> file contains the subspecies assignments.</p>
<p>The visualisation command is a bit more involved as we need to point to both directories and the model directory:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb21"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="ex">poppunk_visualise</span> <span class="at">--ref-db</span> b_uniformis_db <span class="at">--query-db</span> b_uniformis_queries <span class="at">--model-dir</span> b_uniformis_threshold <span class="at">--microreact</span> <span class="at">--output</span> b_uniformis_query_viz <span class="at">--threads</span> 4 <span class="at">--maxIter</span> 10000000 <span class="at">--previous-query-clustering</span> b_uniformis_queries/b_uniformis_queries_clusters.csv <span class="at">--previous-clustering</span> b_uniformis_threshold/b_uniformis_threshold_clusters.csv</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>Load the <code>.microreact</code> output file in the <code>b_uniformis_query_viz</code> directory in <a href="https://microreact.org/">Microreact</a> again to see the output.</p>
<p>Use the menu under the â€˜eyeâ€™ to change from reference/query colour (â€˜Statusâ€™) to â€˜Cluster_Clusterâ€™ to see the clusters.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Challenge
</div>
</div>
<div class="callout-body-container callout-body">
<p>Where do the new genomes sit on the tree relative to the old ones? Do they form a new clade, or sit across existing clades?</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center collapsed" data-bs-toggle="collapse" data-bs-target=".callout-15-contents" aria-controls="callout-15" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Note</span>Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-15" class="callout-15-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Youâ€™ll see that some MAGs formed new clusters (â€˜novel clustersâ€™), whilst others were merged, as the new genomes connected to multiple clusters generated in the original database.</p>
</div>
</div>
</div>
</section>
<section id="updating-the-database" class="level3">
<h3 class="anchored" data-anchor-id="updating-the-database">Updating the database</h3>
<p>It is possible to permanently add the query genomes to the database, so future users can make use of novel cluster assignments. Simply add <code>--update-db</code> to the assignment command above. This is beyond the scope of this practical but is documented <a href="https://poppunk-docs.bacpop.org/query_assignment.html#updating-the-database">here</a> and <a href="https://poppunk-docs.bacpop.org/model_distribution.html">here</a>.</p>
</section>
<section id="further-analysis-1" class="level3">
<h3 class="anchored" data-anchor-id="further-analysis-1">Further analysis</h3>
<p>If you have time, try generating another visualisation of the query network using Cytoscape.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Challenge
</div>
</div>
<div class="callout-body-container callout-body">
<p>How does this network compare to the original one from the previous section, when you only built using isolate genomes and not MAG data?</p>
</div>
</div>
<p>Also, try building a PopPUNK model as before, but only using the MAG data you used from assignment.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Challenge
</div>
</div>
<div class="callout-body-container callout-body">
<p>How do the distance distributions compare between the isolate data and MAG data? Does a threshold model work with the data, or would an automated method for model fitting work better?</p>
</div>
</div>
</section>
</section>
<section id="finding-core-and-accessory-genes" class="level2">
<h2 class="anchored" data-anchor-id="finding-core-and-accessory-genes">3. Finding core and accessory genes</h2>
<p>We can use a probabilistic model to correct for the fact that we know MAGs are missing some genes. If we donâ€™t do this then we will systematically under-estimate their population frequency, and end up with nothing in the core genome.</p>
<section id="celebrimbor-a-pipeline-for-pangenome-analysis-and-threshold-correction" class="level3">
<h3 class="anchored" data-anchor-id="celebrimbor-a-pipeline-for-pangenome-analysis-and-threshold-correction">CELEBRIMBOR: a pipeline for pangenome analysis and threshold correction</h3>
<p>The CELEBRIMBOR prerequisities are already install on the VM. The source code files can be found in the <code>CELEBRIMBOR</code> directory which can be listed here:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb22"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> ~/CELEBRIMBOR-1.1.2</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>As well be editing the config files, make a copy in your working directory:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb23"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> <span class="at">-r</span> ~/CELEBRIMBOR-1.1.2 ~/course_dir/work_dir/Day_5/pangenome/CELEBRIMBOR</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>The files we will be analysing can be listed here:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb24"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> ~/course_dir/data_dir/MGnify_training_course/MAGs_for_CELEBRIMBOR</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>This archive contains two directories. <code>CELEBRIMBOR_MAGs</code> contains the fasta files we will be analysing. <code>results</code> contains the <a href="https://github.com/oschwengers/bakta">Bakta</a> and <a href="https://github.com/Ecogenomics/CheckM">CheckM</a> outputs which were generated previously from these genomes. As these are the slowest parts of the analysis, we have provided them to allow you to generate results faster.</p>
<p>As weâ€™ll be updating <code>results</code>, copy the whole directory to your working directory:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb25"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> <span class="at">-r</span> ~/course_dir/data_dir/MGnify_training_course/MAGs_for_CELEBRIMBOR/results ~/course_dir/work_dir/Day_5/pangenome/CELEBRIMBOR_results</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>CELEBRIMBOR is a snakemake pipeline which enables automated gene annotation, clustering, completeness estimation and core threshold adjustment. Snakemake allows workflows to be re-run if workflows stop prematurely.</p>
<p>To prevent CELEBRIMBOR from re-running Bakta and CheckM, I have edited the source code. If you do your own analysis outside of this practical, you will run the full workflow using identical commands.</p>
<p>Documentation for CELEBRIMBOR can be found <a href="https://github.com/bacpop/CELEBRIMBOR">here</a>.</p>
<p>Snakemake reads a <code>config.yaml</code> file to assign parameters. Navigate to the <code>CELEBRIMBOR</code> directory, and update the <code>config.yaml</code> file with the appropriate parameters.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb26"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/course_dir/work_dir/Day_5/pangenome/CELEBRIMBOR</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="ex">vim</span> config.yaml</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>To start typing in vim, type <code>i</code>. To close vim, first use the <code>esc</code> key, and then WITH saving type, <code>:wq</code>, or WITHOUT saving, type <code>:q!</code></p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb27"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co">#output directory</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="ex">output_dir:</span> /home/training/course_dir/work_dir/Day_5/pangenome/CELEBRIMBOR_results</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="co"># dir with genome fasta files, must have '.fasta' extension, to convert use 'rename .fa .fasta  *.fa' (e.g. if extensions are .fa)</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="ex">genome_fasta:</span> /home/training/course_dir/data_dir/MGnify_training_course/MAGs_for_CELEBRIMBOR/CELEBRIMBOR_MAGs</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a><span class="co"># path to bakta DB:</span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a><span class="ex">bakta_db:</span> /home/training/course_dir/data_dir/db-light</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a><span class="co"># cgt executable parameters</span></span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a><span class="ex">cgt_exe:</span> /home/training/.cargo/bin/cgt_bacpop</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a><span class="ex">cgt_breaks:</span> 0.05,0.95</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a><span class="ex">cgt_error:</span> 0.05</span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a><span class="co"># choice of clustering method, either "mmseqs2" or "panaroo"</span></span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a><span class="ex">clustering_method:</span> <span class="st">"mmseqs2"</span></span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a><span class="co"># must be one of "strict", "moderate" or "sensitive"</span></span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a><span class="ex">panaroo_stringency:</span> <span class="st">"strict"</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>Note: paths in the <code>config.yaml</code> file cannot contain <code>~</code> symbols.</p>
<p>To run CELEBRIMBOR, simply run the below command. Snakemake will read the <code>config.yaml</code> file and run CELEBRIMBOR on the associated files, avoiding Bakta and CheckM as these results have already been generated.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb28"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="ex">snakemake</span> <span class="at">--cores</span> 4</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>This will run for a few minutes. Youâ€™ll see a number of tools being run, including <a href="https://www.nature.com/articles/nbt.3988">MMseqs2</a> for clustering, and <a href="https://github.com/bacpop/cgt">cgt</a> for frequency threshold adjustment.</p>
<p>While youâ€™re waiting, feel free to take a look at the <a href="https://academic.oup.com/bioinformatics/article/40/9/btae542/7762100">CELEBRIMBOR</a> or <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6360808/">PopPUNK</a> papers.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Challenge
</div>
</div>
<div class="callout-body-container callout-body">
<p>Once CELEBRIMBOR has finished running, take a look at the outputs printed to the console. What are the new adjusted core and rare thresholds?</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center collapsed" data-bs-toggle="collapse" data-bs-target=".callout-19-contents" aria-controls="callout-19" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Note</span>Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-19" class="callout-19-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Youâ€™ll observe the new frequency thresholds printed to the console.&nbsp;Here, the core threshold was reduced from 95% to 92.73%, whilst the rare threshold was increased from 5% to 10%.</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb29"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Core</span> threshold: <span class="op">&gt;</span>= 102 observations or <span class="op">&gt;</span>= 92.73% frequency</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="ex">Rare</span> threshold: <span class="op">&lt;</span>= 11 observations or <span class="op">&lt;</span>= 10.00% frequency</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</div>
</div>
<p>Take a look at the output files by running:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb30"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/course_dir/work_dir/Day_5/pangenome/CELEBRIMBOR_results <span class="kw">&amp;&amp;</span> <span class="fu">ls</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>The <code>pangenome_summary.tsv</code> describes the assignments of each gene family to a frequency compartment based on the raw frequency values (column order: gene name, gene annotation, frequency, frequency compartment).</p>
<p><code>cgt_output.txt</code> details the adjusted frequency compartment assignments calculated by CELEBRIMBOR (column order: gene name, gene count, adjusted frequency compartment).</p>
<p>Using <code>R</code>, you can plot the different between the adjusted and unadjusted results by updating the script <code>plot_frequency.R</code> with the paths to <code>pangenome_summary.tsv</code> and <code>cgt_output.txt</code>.</p>
<p>Copy the R script to the working directory and edit:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb31"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> ~/course_dir/data_dir/MGnify_training_course/plot_frequency.R ~/course_dir/work_dir/Day_5/pangenome/CELEBRIMBOR_results</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="ex">vim</span> plot_frequency.R</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>Edit the file below, using <code>esc</code> key and <code>dd</code> to delete whole lines if necessary</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb32"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="ex">library</span><span class="er">(</span><span class="ex">ggplot2</span><span class="kw">)</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="co"># read in data</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="ex">pangenome.summary</span> <span class="op">&lt;</span>- read.csv<span class="er">(</span><span class="st">"/home/training/course_dir/work_dir/Day_5/pangenome/CELEBRIMBOR_results/pangenome_summary.tsv"</span><span class="ex">,</span> sep = <span class="st">"\t"</span>, header=FALSE<span class="kw">)</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a><span class="ex">colnames</span><span class="er">(</span><span class="ex">pangenome.summary</span><span class="kw">)</span> <span class="op">&lt;</span>- <span class="ex">c</span><span class="er">(</span><span class="st">"gene_name"</span><span class="ex">,</span> <span class="st">"gene_family"</span>, <span class="st">"frequency"</span>, <span class="st">"compartment_freq"</span><span class="kw">)</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a><span class="ex">pangenome.summary</span><span class="va">$gene_family</span> <span class="op">&lt;</span>- NULL</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a><span class="ex">cgt.summary</span> <span class="op">&lt;</span>- read.csv<span class="er">(</span><span class="st">"/home/training/course_dir/work_dir/Day_5/pangenome/CELEBRIMBOR_results/cgt_output.txt"</span><span class="ex">,</span> sep = <span class="st">"\t"</span>, header=TRUE<span class="kw">)</span></span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a><span class="ex">colnames</span><span class="er">(</span><span class="ex">cgt.summary</span><span class="kw">)</span> <span class="op">&lt;</span>- <span class="ex">c</span><span class="er">(</span><span class="st">"gene_name"</span><span class="ex">,</span> <span class="st">"count"</span>, <span class="st">"compartment_adjusted"</span><span class="kw">)</span></span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge data</span></span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a><span class="ex">total.df</span> <span class="op">&lt;</span>- merge<span class="er">(</span><span class="ex">pangenome.summary,</span> cgt.summary, by = <span class="st">"gene_name"</span><span class="kw">)</span></span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a><span class="co"># stack data</span></span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a><span class="ex">stacked.df</span> <span class="op">&lt;</span>- data.frame<span class="er">(</span><span class="ex">Type</span> = <span class="st">"Unadjusted"</span>, Compartment = total.df<span class="va">$compartment_freq</span><span class="kw">)</span></span>
<span id="cb32-16"><a href="#cb32-16" aria-hidden="true" tabindex="-1"></a><span class="ex">stacked.df</span> <span class="op">&lt;</span>- rbind<span class="er">(</span><span class="ex">stacked.df,</span> data.frame<span class="er">(</span><span class="ex">Type</span> = <span class="st">"Adjusted"</span>, Compartment = total.df<span class="va">$compartment_adjusted</span><span class="kw">))</span></span>
<span id="cb32-17"><a href="#cb32-17" aria-hidden="true" tabindex="-1"></a><span class="ex">stacked.df</span><span class="va">$Compartment</span> <span class="op">&lt;</span>- factor<span class="er">(</span><span class="ex">stacked.df</span><span class="va">$Compartment</span><span class="ex">,</span> levels = c<span class="er">(</span><span class="st">"rare"</span><span class="ex">,</span> <span class="st">"middle"</span>, <span class="st">"core"</span><span class="kw">))</span></span>
<span id="cb32-18"><a href="#cb32-18" aria-hidden="true" tabindex="-1"></a><span class="ex">stacked.df</span><span class="va">$Type</span> <span class="op">&lt;</span>- factor<span class="er">(</span><span class="ex">stacked.df</span><span class="va">$Type</span><span class="ex">,</span> levels = c<span class="er">(</span><span class="st">"Unadjusted"</span><span class="ex">,</span> <span class="st">"Adjusted"</span><span class="kw">))</span></span>
<span id="cb32-19"><a href="#cb32-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-20"><a href="#cb32-20" aria-hidden="true" tabindex="-1"></a><span class="co"># plot data</span></span>
<span id="cb32-21"><a href="#cb32-21" aria-hidden="true" tabindex="-1"></a><span class="ex">p</span> <span class="op">&lt;</span>- ggplot<span class="er">(</span><span class="ex">stacked.df,</span> aes<span class="er">(</span><span class="va">x</span><span class="op">=</span>Type, <span class="ex">fill</span> = Type<span class="kw">))</span> <span class="ex">+</span> facet_grid<span class="er">(</span><span class="ex">Compartment~.,</span> scales = <span class="st">"free_y"</span><span class="kw">)</span> <span class="ex">+</span> geom_bar<span class="er">(</span><span class="kw">)</span> <span class="ex">+</span> xlab<span class="er">(</span><span class="st">"Pangenome analysis type"</span><span class="kw">)</span> <span class="ex">+</span> ylab<span class="er">(</span><span class="st">"Gene count"</span><span class="kw">)</span> <span class="ex">+</span> theme<span class="er">(</span><span class="ex">legend.position</span> = <span class="st">"none"</span><span class="kw">)</span></span>
<span id="cb32-22"><a href="#cb32-22" aria-hidden="true" tabindex="-1"></a><span class="ex">ggsave</span><span class="er">(</span><span class="st">"/home/training/course_dir/work_dir/Day_5/pangenome/CELEBRIMBOR_results/adjusted_gene_frequency.png"</span><span class="ex">,</span> plot = p<span class="kw">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>Run the script from the <code>R</code> terminal:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb33"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"plot_frequency.R"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>Use <code>Ctrl + Z</code> to close the R terminal.</p>
<p>Take a look at <code>adjusted_gene_frequency.png</code> by opening the <code>Files</code> application and navigating to <code>Home -&gt; course_dir -&gt; work_dir -&gt; Day_5 -&gt; pangenome -&gt; CELEBRIMBOR_results</code>.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center collapsed" data-bs-toggle="collapse" data-bs-target=".callout-20-contents" aria-controls="callout-20" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Note</span>Solution
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-20" class="callout-20-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>It should look something like this:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="pangenome/frequency_adjustment.png" class="img-fluid figure-img"></p>
<figcaption>Effect of frequency adjustment</figcaption>
</figure>
</div>
<p>Observe that the core genome and rare genome sizes increase after adjustment, whilst the middle or intermediate decreases.</p>
</div>
</div>
</div>
<p>There are other files that can be used for downstream analysis, such as <code>presence_absence_matrix.txt</code> which defines in which genomes gene families are found, as well as the <code>annotated</code> directory, which contains gene annotations generated using Bakta.</p>
</section>
<section id="further-analysis-2" class="level3">
<h3 class="anchored" data-anchor-id="further-analysis-2">Further analysis</h3>
<p>Try running CELEBRIMBOR with different parameters, such as <code>cgt_breaks</code> which defines the original rare and core thresholds to adjust, or <code>cgt_error</code>, which defines the false negative rate of CELEBRIMBOR. Replot these using the R script and compare to your previous results.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Challenge
</div>
</div>
<div class="callout-body-container callout-body">
<p>What does altering these parameter do to the number of core, middle and rare genes identified?</p>
</div>
</div>
<p>You can also try running with <code>clustering_method</code> set to <code>panaroo</code>, which uses <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02090-4">Panaroo</a>, a more accurate clustering method but less scalable than MMseqs2 (this will likely take a while to run). Also, when running Panaroo, try setting <code>panaroo_stringency</code> to <code>moderate</code> or <code>senstive</code>. Details on the effect of these parameters can be found <a href="https://gthlab.au/panaroo/#/gettingstarted/params">here</a>.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>Challenge
</div>
</div>
<div class="callout-body-container callout-body">
<p>Plot the output of Panaroo against that of MMseqs in R, and see how using different clustering algorithms impacts the number of core, middle and rare genes.</p>
</div>
</div>
</section>
</section>
<section id="example-analysis" class="level2">
<h2 class="anchored" data-anchor-id="example-analysis">Example analysis</h2>
<p>Example analysis for all of todayâ€™s workshop is available here:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb35"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> ~/course_dir/data_dir/MGnify_training_course/example_analysis/pangenome</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>


</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" id="quarto-reuse"><h2 class="anchored quarto-appendix-heading">Reuse</h2><div class="quarto-appendix-contents"><div>Apache 2.0</div></div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "î§‹";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/course25\.mgnify\.org");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../sessions/getting_data_from_mgnify.html" class="pagination-link" aria-label="Getting data from MGnify">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Getting data from MGnify</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav>
</div> <!-- /content -->




</body></html>